File: Versions/js/AudioDBPImporter.js
/**
* This program reads metadata from the DBP server, and generates
* .json files that can be uploaded to S3.
* 1. Manually lookup the DBP language code for each sil language that I have in Versions.db
* 2. The DBP_lang_code should be added to Versions.db Languages table
* 3. Perform a volume list query for each Language, and store the required volume information in flat files.
* 4. Perform a book list query for each damid, and store the required data in flat files.
* 5. Perform a verse position query for each damid/book/chapter, and store the required data in flat files.
* 6. Once this all works well, should I create an option to upload each file as it is created?
*/

"use strict";
var http = require('http');
var file = require('fs');


var audioDBPImporter = function(callback) {
	
	var HOST = "http://dbt.io/";
	var KEY = "key=b37964021bdd346dc602421846bf5683&v=2";
	var DIRECTORY = "output/";
	
	/**
	* This table controls what Audio versions will be included, and what
	* text versions that are associated with
	*/
	var versions = {
			//'ERV-ARB': ['ARB', 'WTC', true ],
			//'ARBVDPD': ['ARB', 'WTC', false ],
			//'ERV-AWA': ['AWA', 'WTC', true ],
			//'ERV-BEN': ['BNG', 'WTC', true ],
			'ERV-BUL': ['BLG', 'AMB', false ],
			'ERV-CMN': ['CHN', 'UNV', true ], // for mainland China
			//'ERV-CMN': ['YUH', 'UNV', false ], // for Hong Kong
			'ERV-ENG': ['ENG', 'ESV', false ], // must change to WEB in production
			'KJVPD':   ['ENG', 'KJV', true ],
			'WEB':     ['ENG', 'WEB', true ],
			'ERV-HRV': ['SRC', null, false ],
			//'ESV':	   ['ENG', 'ESV', true],
			'ERV-HIN': ['HIN', null, false],
			'ERV-HUN': ['HUN', 'HBS', false ],
			'ERV-IND': ['INZ', 'SHL', false ],
			//'ERV-KAN': ['ERV', 'WTC', true ],
			'ERV-MAR': ['MAR', null, false ],
			'ERV-NEP': ['NEP', null, false ],
			'ERV-ORI': ['ORY', null, false ],
			'ERV-PAN': ['PAN', null, false ],
			'ERV-POR': ['POR', 'ARA', false ],
			'ERV-RUS': ['RUS', 'S76', false ],
			//'ERV-SPA': ['SPN', 'WTC', true ],
			'ERV-SPA': ['SPN', 'R95', false ], // or R95 or BDA
			'ERV-SRP': ['SRP', null, false ],
			//'ERV-TAM': ['TCV', 'WTC', true ],
			'ERV-THA': ['THA', null, false ],
			'ERV-UKR': ['UKR', 'O95', false ],
			//'ERV-URD': ['URD', 'WTC', true ],
			'ERV-URD': ['URD', 'PAK', false ],
			'ERV-VIE': ['VIE', null, false ],
			'NMV':     ['PES', null, false ]
	};
	
//	var versions = {
//		'ERV-ENG': ['ENG', 'ESV', false ]	
//	};
	
	var versionList = Object.getOwnPropertyNames(versions);
	console.log("VERS " + versionList);
	var damIdList = [];
	var bookList = [];
	var audioVersionSql = [];
	var audioTableSql = [];
	var audioBookTableSql = [];
	var audioChapterTableSql = [];
	
	doAllVersions(versionList, function() {
		//console.log(damIdList);
		doAllVolumes(damIdList, function() {
			//console.log(bookList);
			doAllChapters(bookList, function() {
				callback();
			});
		});
	});
	
	function doAllVersions(versionList, callback) {
		var version = versionList.shift();
		if (version) {
			var DBPData = versions[version];
			var dbpLanguage = DBPData[0];
			var dbpVersion = DBPData[1];
			if (dbpVersion) {
				var sql = "INSERT INTO AudioVersion VALUES ('" + version + "', '" + 
				dbpLanguage + "', '" + dbpVersion + "');";
				console.log(sql);
				audioVersionSql.push(sql);
			}
			doVolumeListQuery(dbpLanguage, dbpVersion, function() {
				doAllVersions(versionList, callback);
			});
		} else {
			writeSQLFile(DIRECTORY + 'AudioVersionTable.sql', audioVersionSql);
			
			writeSQLFile(DIRECTORY + 'AudioTable.sql', audioTableSql);
				
			callback();	
		}
	}
	
	function doVolumeListQuery(dbpLanguage, dbpVersion, callback) {
		var url = HOST + "library/volume?" + KEY + "&media=audio&language_code=" + dbpLanguage;
		httpGet(url, function(json) {
			console.log("Before Prune " + json.length);
			//console.log(json);
			var jsonVersion = pruneListByVersion(json, dbpVersion);
			console.log("After Prune " + jsonVersion.length);			
			console.log(jsonVersion);
			var nameList = [ 'dam_id', 'language_code', 'version_code', 'collection_code', 'media_type', 'volume_name' ];
			var sqlResult = insertStmt('Audio', nameList, jsonVersion);
			for (var r=0; r<sqlResult.length; r++) {
				var row = sqlResult[r];
				var date = new Date().toISOString().substr(0, 10);
				row.splice(row.length - 1, 0, ", '" + date + "'");
				console.log(row.join(''));
				audioTableSql.push(row.join(''));
			}
			callback();
		});
	}
	
	function pruneListByVersion(volumeList, versionCode) {
		var result = [];
		for (var i=0; i<volumeList.length; i++) {
			var volume = volumeList[i];
			if (volume.version_code == versionCode) {
				result.push(volume);
				damIdList.push(volume.dam_id);
				
			}
		}
		return(result);
	}
	
	function doAllVolumes(damIdList, callback) {
		var damId = damIdList.shift();
		if (damId) {
			doBookListQuery(damId, function() {
				doAllVolumes(damIdList, callback);
			});
		} else {
			writeSQLFile(DIRECTORY + 'AudioBookTable.sql', audioBookTableSql);
			callback();
		}
	}

	function doBookListQuery(damId, callback) {
		var url = HOST + "library/book?" + KEY + "&dam_id=" + damId;
		httpGet(url, function(json) {
			console.log(json.length);
			for (var i=0; i<json.length; i++) {
				var result = json[i];
				if (result.book_order.length == 1) {
					result.book_order = "0" + result.book_order;
				}			
				result.usfm_book_id = getUSFMBookCode(result.book_id);
				bookList.push(result);
			}
			var nameList = ['dam_id', 'usfm_book_id', 'book_order', 'number_of_chapters'];
			var sqlResult = insertStmt('AudioBook', nameList, json);

			for (i=0; i<sqlResult.length; i++) {
				var row = sqlResult[i];
				console.log(row.join(''));
				audioBookTableSql.push(row.join(''));
			}
			callback();
		});
	}
	
	function doAllChapters(bookList, callback) {
		var book = bookList.shift();
		if (book) {
			console.log('Verses ' + book.dam_id + '  ' + book.book_id);
			var numOfChapters = 0 + book.number_of_chapters;
			doEachChapter(book, 1, numOfChapters, function() {
				doAllChapters(bookList, callback);
			});
		} else {
			writeSQLFile(DIRECTORY + 'AudioChapterTable.sql', audioChapterTableSql);
			callback();
		}
	}
	
	function doEachChapter(book, chapterNum, numOfChapters, callback) {
		if (chapterNum <= numOfChapters) {
			doVerseListQuery(book, chapterNum, function(count) {
				if (chapterNum == 1 && count == 0) {
					callback();
					return;
				}
				doEachChapter(book, chapterNum + 1, numOfChapters, callback);
			});
		} else {	
			callback();
		}
	}
	
	function doVerseListQuery(book, chapterNum, callback) {
		var url = HOST + "audio/versestart?" + KEY + "&dam_id=" + book.dam_id + "&osis_code=" + book.book_id + "&chapter_number=" + chapterNum;
		httpGet(url, function(json) {
			var row = {};
			for (var i=0; i<json.length; i++) {
				var item = json[i];
				row[item.verse_id] = item.verse_start;
			}
			if (json.length > 0) {
				var sql = "INSERT INTO AudioChapter VALUES('" + book.dam_id + "', '" + book.usfm_book_id + "', '" + chapterNum + "', '" + JSON.stringify(row) + "');"
				console.log(sql);
				audioChapterTableSql.push(sql);
			}
			callback(json.length);	
		});	
	}
	
	/**
	* Return the USFM book code when given the OSIS book code
	*/
	function getUSFMBookCode(bookCode) {
		var books = {
			'Gen':   'GEN',
			'Exod':  'EXO',
			'Lev':   'LEV',
			'Num':   'NUM',
			'Deut':  'DEU',
			'Josh':  'JOS',
			'Judg':  'JDG',
			'Ruth':  'RUT',
			'1Sam':  '1SA',
			'2Sam':  '2SA',
			'1Kgs':  '1KI',
			'2Kgs':  '2KI',
			'1Chr':  '1CH',
			'2Chr':  '2CH',
			'Ezra':  'EZR',
			'Neh':   'NEH',
			'Esth':  'EST',
			'Job':   'JOB',
			'Ps':    'PSA',
			'Prov':  'PRO',
			'Eccl':  'ECC',
			'Song':  'SNG',
			'Isa':   'ISA',
			'Jer':   'JER',
			'Lam':   'LAM',
			'Ezek':  'EZK',
			'Dan':   'DAN',
			'Hos':   'HOS',
			'Joel':  'JOL',
			'Amos':  'AMO',
			'Obad':  'OBA',
			'Jonah': 'JON',
			'Mic':   'MIC',
			'Nah':   'NAM',
			'Hab':   'HAB',
			'Zeph':  'ZEP',
			'Hag':   'HAG',
			'Zech':  'ZEC',
			'Mal':   'MAL',
			'Matt':  'MAT',
			'Mark':  'MRK',
			'Luke':  'LUK',
			'John':  'JHN',
			'Acts':  'ACT',
			'Rom':   'ROM',
			'1Cor':  '1CO',
			'2Cor':  '2CO',
			'Gal':   'GAL',
			'Eph':   'EPH',
			'Phil':  'PHP',
			'Col':   'COL',
			'1Thess':'1TH',
			'2Thess':'2TH',
			'1Tim':  '1TI',
			'2Tim':  '2TI',
			'Titus': 'TIT',
			'Phlm':  'PHM',
			'Heb':   'HEB',
			'Jas':   'JAS',
			'1Pet':  '1PE',
			'2Pet':  '2PE',
			'1John': '1JN',
			'2John': '2JN',
			'3John': '3JN',
			'Jude':  'JUD',
			'Rev':   'REV'
		};
		var result = books[bookCode];
		return((result) ? result : bookCode);
	}
	
	function httpGet(url, callback) {
		http.get(url, function(response) {
		  	if (response.statusCode !== 200) {
		    	var error = new Error('Request Failed. + Status Code: ' + response.statusCode);
				response.resume();
				errorMessage(error, url);
				return;
		  	}
		
		  	response.setEncoding('utf8');
		  	var rawData = '';
		  	response.on('data', function(chunk) { rawData += chunk; });
		  	response.on('end', function() {
		    	try {
		      		callback(JSON.parse(rawData));
		    	} catch (e) {
		      		errorMessage(e, url);
		    	}
		  	});
		}).on('error', function(e) {
		  errorMessage(e, url);
		});
	}
	
	function insertStmt(table, columnList, json) {
		var array = [];
		for (var i=0; i<json.length; i++) {
			var row = [];
			var item = json[i];
			row.push("REPLACE INTO " + table + " VALUES (");
			for (var n=0; n<columnList.length; n++) {
				if (n > 0) {
					row.push(', ');
				}
				var col = columnList[n];
				row.push("'");
				row.push(item[col]);
				row.push("'");
			}
			row.push(");");
			array.push(row);
		}
		return array;
	}
	
	function ensureDirectory(directory) {
		if (!file.existsSync(directory)) {
			file.mkdirSync(directory);
		}
	}
	
	function writeSQLFile(filename, array) {
		file.writeFile(filename, array.join('\n'), 'utf8', function(err) {
			if (err) {
				errorMessage(err, filename);
			}
		});
	}
	
	function errorMessage(error, message) {
		console.log('ERROR', message, JSON.stringify(error.message));
		process.exit(1);
	}
};


audioDBPImporter(function() {
	console.log('DONE WITH CREATE META DATA');
});

File: Versions/py/AudioDBPImporter.py
#
# The purpose of this program is to validate the book code
# associated with each audio bible in dbp-prod
#

# TO Do
# 1. Write a program that parses each bookname, and looks it up in this table to report missing names
# 2. Also build a table of Book Order associated with each BookName or usfm bookCode if possible
# 3. Output any name mismatches
# 4. Output order codes for usfm bookCodes, and for bookNames
# 5. Put the resulting corresponding usfm -> [Order Code, Name]

import io
import sys

# This table controls what Audio versions will be included, and what
# text versions that are associated with
versions = [
	['ERV-ARB', 'ARB', 'WTC', True ], 	# ARBWTC/ARBWTCN1DA, ARBWTCO1DA
	['ARBVDPD', 'ARB', 'VDV', True ], 	# ARBVDV/ARZVDVN2DA, ARZVDVO2DA
	['ERV-AWA', 'AWA', 'WTC', True ], 	# AWAWTC/AWAWTCN2DA,
	['ERV-BEN', 'BEN', 'WTC', True ], 	# BENWTC/BNGWTCN1DA, BNGWTCN2DA
	['ERV-BUL', 'BUL', 'PRB', False ], 	# BULPRB/BLGAMBN1DA
	['ERV-CMN', 'CMN', 'UN1', True ], 	# CMNUN1/CHNUNVN2DA, CMNUNV/CHNUNVO2DA
	['ERV-ENG', 'ENG', 'ESV', False ], 	# ENGESV/ENGESVN2DA, ENGESVO2DA
	['KJVPD', 	'ENG', 'KJV', True ], 	# ENGKJV/ENGKJVN2DA, ENGKJVO2DA
	['WEB', 	'ENG', 'WEB', True ], 	# ENGWEB/ENGWEBN2DA, ENGWEBO2DA
	['ERV-HRV', 'SRC', None, False ],
	['ERV-HIN', 'HIN', 'WTC', True], 	# HINWTC/HNDWTCN2DA
	['ERV-HUN', 'HUN', 'HBS', False ],	# HUNHBS/HUNHBSN1DA
	['ERV-IND', 'IND', 'SHL', False ],	# INDSHL/INZSHLN2DA
	['ERV-KAN', 'KAN', 'WTC', True ],	# KANWTC/ERVWTCN1DA, ERVWTCN2DA
	['ERV-MAR', 'MAR', 'WTC', True ],	# MARWTC/MARWTCN1DA, MARWTCN2DA
	['ERV-NEP', 'NEP', None, False ],
	['ERV-ORI', 'ORY', 'WTC', True ],	# ORYWTC/ORYWTCN1DA, ORYWTCN2DA
	['ERV-PAN', 'PAN', None, False ],
	['ERV-POR', 'POR', 'BAR', False ],	# PORBAR/PORARAN2DA
	['ERV-RUS', 'RUS', 'S76', False ],	# RUSS76/RUSS76N2DA, RUSS76O2DA
	['ERV-SPA', 'SPA', 'WTC', True ],	# SPAWTC/SPNWTCN2DA
	['ERV-SRP', 'SRP', None, False ],
	['ERV-TAM', 'TAM', 'WTC', True ],	# TAMWTC/TCVWTCN1DA
	['ERV-THA', 'THA', None, False ],
	['ERV-UKR', 'UKR', 'N39', False ],	# UKRN39/UKRO95N2DA
	['ERV-URD', 'URD', 'WTC', True ],	# URDWTC/URDWTCN2DA
	['ERV-VIE', 'VIE', None, False ],
	['NMV', 	'PES', None, False ]
]

#for version in versions:
#	print version

def usfmBookId(bookName):
	books = {
		'Genesis':   	'GEN',
		'Exodus':  		'EXO',
		'Leviticus':   	'LEV',
		'Numbers':   	'NUM',
		'Deuteronomy':  'DEU',
		'Joshua':  		'JOS',
		'Judges':  		'JDG',
		'Ruth':  		'RUT',
		'1Samuel':  	'1SA',
		'2Samuel':  	'2SA',
		'1Kings':  		'1KI',
		'2Kings':  		'2KI',
		'1Chronicles':  '1CH',
		'2Chronicles':  '2CH',
		'Ezra':  		'EZR',
		'Nehemiah':   	'NEH',
		'Esther':  		'EST',
		'Job':   		'JOB',
		'Psalms':    	'PSA',
		'Proverbs':  	'PRO',
		'Ecclesiastes': 'ECC',
		'SongofSongs':  'SNG',
		'Isaiah':   	'ISA',
		'Jeremiah':   	'JER',
		'Lamentations': 'LAM',
		'Ezekiel':  	'EZK',
		'Daniel':   	'DAN',
		'Hosea':   		'HOS',
		'Joel':  		'JOL',
		'Amos':  		'AMO',
		'Obadiah':  	'OBA',
		'Jonah': 		'JON',
		'Micah':   		'MIC',
		'Nahum':   		'NAM',
		'Habakkuk':   	'HAB',
		'Zephaniah':  	'ZEP',
		'Haggai':   	'HAG',
		'Zechariah':  	'ZEC',
		'Malachi':   	'MAL',
		'Matthew':  	'MAT',
		'Mark':  		'MRK',
		'Luke':  		'LUK',
		'John':  		'JHN',
		'Acts':  		'ACT',
		'Romans':   	'ROM',
		'1Corinthians': '1CO',
		'2Corinthians': '2CO',
		'Galatians':   	'GAL',
		'Ephesians':   	'EPH',
		'Philippians':  'PHP',
		'Colossians':   'COL',
		'1Thess':		'1TH',
		'2Thess':		'2TH',
		'1Timothy':  	'1TI',
		'2Timothy':  	'2TI',
		'Titus': 		'TIT',
		'Philemon':  	'PHM',
		'Hebrews':   	'HEB',
		'James':   		'JAS',
		'1Peter':  		'1PE',
		'2Peter':  		'2PE',
		'1John': 		'1JN',
		'2John': 		'2JN',
		'3John': 		'3JN',
		'Jude':  		'JUD',
		'Revelation':   'REV',
		# Spanish
		'San Mateo':	'MAT',
		'San Marcos':	'MRK',
		'San Lucas':	'LUK',
		'San Juan':		'JHN',
		'Hechos':		'ACT',
		'Romanos':		'ROM',
		'1Corintios':	'1CO',
		'2Corintios':	'2CO',
		'Galatas':		'GAL',
		'Efesios':		'EPH',
		'Filipenses':	'PHP',
		'Colosenses':	'COL',
		'1Tes':			'1TH',
		'2Tes':			'2TH',
		'1Timoteo':		'1TI',
		'2Timoteo':		'2TI',
		'Tito':			'TIT',
		'Filemon':		'PHM',
		'Hebreos':		'HEB',
		'Santiago':		'JAS',
		'1San Pedro':	'1PE',
		'2San Pedro':	'2PE',
		'1San Juan':	'1JN',
		'2San Juan':	'2JN',
		'3San Juan':	'3JN',
		'Judas':		'JUD',
		'Apocalipsis':	'REV',
		# Portuguese
		'S Mateus':		'MAT',
		'S Marcos':		'MRK',
		'S Lucas':		'LUK',
		'S Joao':		'JHN',
		'Atos':			'ACT',
		'Colossenses':	'COL',
		'1Tess':		'1TH',
		'2Tess':		'2TH',
		'Hebreus':		'HEB',
		'S Tiago':		'JAS',
		'1Pedro':		'1PE',
		'2Pedro':		'2PE',
		'1S Joao':		'1JN',
		'2S Joao':		'2JN',
		'3S Joao':		'3JN',
		'S Judas':		'JUD',
		'Apocalipse':	'REV',
		# Indonesian
		'Matius':		'MAT',
		'Markus':		'MRK',
		'Lukas':		'LUK',
		'Yohanes':		'JHN',
		'Kisah Rasul':	'ACT',
		'Roma':			'ROM',
		'1Korintus':	'1CO',
		'2Korintus':	'2CO',
		'Galatia':		'GAL',
		'Efesus':		'EPH',
		'Filipi':		'PHP',
		'Kolose':		'COL',
		'1Tesalonika':	'1TH',
		'2Tesalonika':	'2TH',
		'1Timotius':	'1TI',
		'2Timotius':	'2TI',
		'Ibrani':		'HEB',
		'Yakobus':		'JAS',
		'1Petrus':		'1PE',
		'2Petrus':		'2PE',
		'1Yohanes':		'1JN',
		'2Yohanes':		'2JN',
		'3Yohanes':		'3JN',
		'Yudas':		'JUD',
		'Wahyu':		'REV'
	}
	result = books.get(bookName, None)
	return result

underscores = "_______________"

abbrDict = dict()	
for version in versions:
	if version[2] != None:
		abbr = version[1] + version[2]
		abbrDict[abbr] = version[0]

versionOut = io.open("output/AudioVersionTable.sql", mode="w", encoding="utf-8")
versionOut.write(u"DROP TABLE IF EXISTS AudioVersion;\n")
versionOut.write(u"CREATE TABLE AudioVersion(\n")
versionOut.write(u"  ssVersionCode TEXT NOT NULL PRIMARY KEY,\n")
versionOut.write(u"  dbpLanguageCode TEXT NOT NULL,\n")
versionOut.write(u"  dbpVersionCode TEXT NOT NULL);\n")

audioOut = io.open("output/AudioTable.sql", mode="w", encoding="utf-8")
audioOut.write(u"DROP TABLE IF EXISTS Audio;\n")
audioOut.write(u"CREATE TABLE Audio(\n")
audioOut.write(u"  damId TEXT NOT NULL PRIMARY KEY,\n")
audioOut.write(u"  dbpLanguageCode TEXT NOT NULL,\n")
audioOut.write(u"  dbpVersionCode TEXT NOT NULL,\n")
audioOut.write(u"  collectionCode TEXT NOT NULL,\n")
audioOut.write(u"  mediaType TEXT NOT NULL);\n")

bookOut = io.open("output/AudioBookTable.sql", mode="w", encoding="utf-8")
bookOut.write(u"DROP TABLE IF EXISTS AudioBook;\n")
bookOut.write(u"CREATE TABLE AudioBook(\n")
bookOut.write(u"  damId TEXT NOT NULL REFERENCES Audio(damId),\n")
bookOut.write(u"  bookId TEXT NOT NULL,\n")
bookOut.write(u"  bookOrder TEXT NOT NULL,\n")
bookOut.write(u"  bookName TEXT NOT NULL,\n")
bookOut.write(u"  numberOfChapters INTEGER NOT NULL,\n")
bookOut.write(u"  PRIMARY KEY (damId, bookId));\n")


versionIdSet = set()
damIdSet = set()
lastDamId = None
lastUsfm = None
bookLine = None

dbpProd = io.open("Release.1.13/metadata/FCBH/dbp_prod.txt", mode="r", encoding="utf-8")
for line in dbpProd:
	line = line.strip()
	parts = line.split("/")
	numParts = len(parts)
	if parts[0] == 'audio' and parts[numParts -1][-4:] == ".mp3":
		abbr = parts[1]
		damId = parts[2]
		if len(abbr) == 6 and len(damId) == 10 and numParts == 4 and abbr in abbrDict.keys():
			if not abbr in versionIdSet:
				versionIdSet.add(abbr)
				ssVersionCode = abbrDict[abbr]
				versionOut.write(u"INSERT INTO AudioVersion VALUES ('%s', '%s', '%s');\n"
				% (ssVersionCode, abbr[0:3], abbr[3:]))
			book = parts[3]
			testament = book[0:1]
			if testament == 'A' or testament == 'B':
				order = book[0:3]
				chapter = book[5:8]
				chapter = chapter.replace("_", "")
				name = book[9:21]
				name = name.replace("_", " ").strip()
				damId2 = book[21:31].replace("_", " ").strip()
				if damId == damId2:
					usfm = usfmBookId(name)
					if usfm == None:
						print "ERROR", line, name
					checkChapter = chapter
					if len(checkChapter) < 3:
						checkChapter = "_" + checkChapter
					checkName = name.replace(" ", "_")
					checkName = checkName + underscores[0: 12 - len(name)]
					generated = "audio/%s/%s/%s__%s_%s%s.mp3" % (abbr, damId, order, checkChapter, checkName, damId)
					if line != generated:
						print "ERROR"
						print line
						print generated
					collectionCode = damId[6:7] + "T"
					if collectionCode == 'OT' or collectionCode == 'NT':
						mType = damId[7:]
						if mType != '1DA' and mType != '2DA':
							print "ERROR mediaType", line
						mediaType = 'Drama' if (mType == '2DA') else 'Non-Drama'
						if not damId in damIdSet:
							damIdSet.add(damId)
							audioOut.write(u"REPLACE INTO Audio VALUES('%s', '%s', '%s', '%s', '%s');\n"
							% (damId, abbr[0:3], abbr[3:6], collectionCode, mediaType))
						bookIdKey = damId + usfm
						if usfm != lastUsfm or damId != lastDamId:
							if bookLine != None:
								bookOut.write(bookLine)
							lastUsfm = usfm
							lastDamId = damId
						bookLine = u"REPLACE INTO AudioBook VALUES('%s', '%s', '%s', '%s', '%s');\n" % (damId, usfm, order, name, chapter)


dbpProd.close()
versionOut.close()
audioOut.close()
bookOut.write(bookLine)
bookOut.close()

#for order in orderSet:
#	print order

#for chapter in chapterSet.keys():
#	print chapter, chapterSet[chapter]

#for book in bookSet:
#	print book
